near-sdk-macros/src/lib.rs

```diff
@@ -108,8 +138,37 @@ pub fn near_bindgen(attr: TokenStream, item: TokenStream) -> TokenStream {
             #input
             #ext_gen
             #abi_embedded
+            #metadata
         })
     } else if let Ok(mut input) = syn::parse::<ItemImpl>(item) {
+        for method in input.items.iter() {
+            if let ImplItem::Fn(m) = method {
+                let ident = &m.sig.ident;
+                if ident.eq("__contract_abi") || ident.eq("contract_source_metadata") {
+                    return TokenStream::from(
+                        syn::Error::new_spanned(
+                            m.sig.ident.to_token_stream(),
+                            "use of reserved contract method",
+                        )
+                        .to_compile_error(),
+                    );
+                }
+            }
+        }
+
+        if input.trait_.is_none() {
+            match syn::parse_str::<syn::ImplItem>(
+                "pub fn contract_source_metadata() -> String { CONTRACT_SOURCE_METADATA.to_string() }",
```

Comment:
It should return a JSON object, not a JSON object serialized to JSON string, so the return value has to be a unit type and then you can use `env::value_return()`:

```rust
                    pub fn contract_source_metadata() {
                        env::value_return(CONTRACT_SOURCE_METADATA);
```